### PR TITLE
Persist unknown build ids

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
- Persist unknown build ids in versioning data when tasks are added
- Add metrics for unknown build ids for add and poll

**Why?**
This (combined with future PRs) will fix a few situations with replication lag, propagation lag, and will also enable cross-task-queue activities, child workflows, and CAN that are versioned as a bundle.

**How did you test it?**
new unit + integration tests

### CODE REVIEW

File: service/matching/db.go [hunk 1]
```diff
@@ -352,7 +352,7 @@ func (db *taskQueueDB) loadUserData(ctx context.Context) error {
 
 	response, err := db.store.GetTaskQueueUserData(ctx, &persistence.GetTaskQueueUserDataRequest{
 		NamespaceID: db.namespaceID.String(),
-		TaskQueue:   db.taskQueue.BaseNameString(),
+		TaskQueue:   db.taskQueue.FullName(),
```
Comment (MichaelSnowden) [inline service/matching/db.go:355]: Why was this changed?

File: service/matching/matching_engine_test.go [hunk 2]
```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
...
+
+	_, err := s.matchingEngine.getTask(ctx, tlID, normalStickyInfo, &pollMetadata{
+		workerVersionCapabilities: &commonpb.WorkerVersionCapabilities{
+			BuildId:       "unknown",
+			UseVersioning: true,
+		},
+	})
+	s.Error(err) // deadline exceeded or canceled
+
+	unknownCtr := scope.Snapshot().Counters()["test.unknown_build_polls+namespace="+matchingTestNamespace+",operation=TaskQueueMgr,task_type=Workflow,taskqueue=makeToast"]
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:2344]: Not for this PR, but I think we should have a better API for querying this. It's not great for it to be both:
1. Sensitive to the ordering of the tags
2. Require you to specify all tags--even ones you may not necessarily care about

File: service/matching/matching_engine_test.go [hunk 3]
```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
...
+		VersionDirective: &taskqueue.TaskVersionDirective{
+			Value: &taskqueue.TaskVersionDirective_BuildId{
+				BuildId: "unknown",
+			},
+		},
+	})
+	s.ErrorIs(err, errRemoteSyncMatchFailed)
+
+	unknownCtr := scope.Snapshot().Counters()["test.unknown_build_tasks+namespace="+matchingTestNamespace+",operation=TaskQueueMgr,task_type=Workflow,taskqueue=makeToast"]
+	s.Equal(int64(1), unknownCtr.Value())
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:2382]: It's really good that we're asserting these oft-forgotten things like error types and metrics. However, shouldn't there also be more critical side effects that we're verifying? For example, shouldn't we verify that something is or is not persisted? Am I right in understanding that there is nothing persisted here due to the error, and that is implicitly asserted due to the absence of an EXPECT() statement?

File: service/matching/matching_engine_test.go [hunk 4]
```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
...
+			TaskQueue:              &taskqueuepb.TaskQueue{Name: tl, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+			ScheduleToStartTimeout: timestamp.DurationFromSeconds(100),
+			// do not set ForwardedSource, allow to go to db
+			VersionDirective: &taskqueue.TaskVersionDirective{
+				Value: &taskqueue.TaskVersionDirective_BuildId{
+					BuildId: "unknown",
+				},
+			},
+		})
+		s.NoError(err)
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:2420]: Can you verify that this works from another goroutine? I think there's some weird behavior of the FailNow() family of methods when called in the background.

File: service/matching/matching_engine_test.go [hunk 5]
```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
...
+		wg.Done()
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:2421]: Will this statement be reached if 2420 fails?

File: service/matching/matching_engine_test.go [hunk 6]
```diff
@@ -2322,6 +2322,124 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
...
+	}()
+
+	go func() {
+		time.Sleep(10 * time.Millisecond)
```
Comment (MichaelSnowden) [inline service/matching/matching_engine_test.go:2425]: Could you comment why we need a sleep here? Also, could you run with a high `-count` to verify that this isn't flaky? Not that we have to eradicate time.Sleep from all tests; it might be prohibitively difficult to avoid in this case, but I want to at least make sure it isn't flaky.

File: service/matching/task_queue_manager.go [hunk 7]
```diff
@@ -809,20 +818,52 @@ func (c *taskQueueManagerImpl) RedirectToVersionedQueueForAdd(ctx context.Contex
...
-		err := checkVersionForStickyAdd(data, buildId)
-		return c.taskQueueID, userDataChanged, err
+		unknownBuild, err := checkVersionForStickyAdd(data, buildId)
+		if err != nil {
+			return nil, nil, err
+		}
+		if unknownBuild {
+			c.recordUnknownBuildTask(buildId)
+			// Don't bother persisting the unknown build id in this case: sticky tasks have a
+			// short timeout, so it doesn't matter if they get lost.
```
Comment (MichaelSnowden) [inline service/matching/task_queue_manager.go:828]: Why does sticky tasks having a short timeout imply that it doesn't matter if they're dropped?

File: service/matching/task_queue_manager.go [hunk 8]
```diff
@@ -809,20 +818,52 @@ func (c *taskQueueManagerImpl) RedirectToVersionedQueueForAdd(ctx context.Contex
...
+		}
+		return c.taskQueueID, userDataChanged, nil
 	}
 
-	versionSet, err := lookupVersionSetForAdd(data, buildId)
+	versionSet, unknownBuild, err := lookupVersionSetForAdd(data, buildId)
```
Comment (MichaelSnowden) [inline service/matching/task_queue_manager.go:833]: I'm worried that if there's another non-retryable error like an unknown build ID here, then the number of branches here will just explode. Is there a way we can just stuff the unknownBuild flag into the error type here and treat it the same as any other non-retryable error?

File: service/matching/task_queue_manager.go (hunk 7 above)
Comment (dnr) [reply service/matching/task_queue_manager.go:828]: Because history will retry on the normal queue is 5 seconds anyway.

The reason we're persisting is because of this unlikely sequence:
1. Matching gets task with unknown build id; it guesses a version set id for it
2. There's no poller, so it writes it to the db
3. That task queue gets unloaded due to idleness
4. The replication finishes and it turns out that build id actually belonged to another set
5. Now a poller comes for it

If we don't persist the build id, the specific versioned task queue will be lost to that poller, since any new task or poll will find the other set for it. So for tasks with no timeout, that we're not allowed to drop, we need to persist so we can find it again. But for sticky tasks with a 5s timeout, the chances that all the above happens within 5s is extraordinarily unlikely, and even if it did happen, dropping the task is totally fine and easier

File: service/matching/task_queue_manager.go (hunk 8 above)
Comment (dnr) [reply service/matching/task_queue_manager.go:833]: unknownBuild being true isn't an error, though. If I return both a value and an "error", won't you complain that that's confusing not idiomatic?

File: service/matching/task_queue_manager.go [hunk 9]
```diff
@@ -809,20 +818,52 @@ func (c *taskQueueManagerImpl) RedirectToVersionedQueueForAdd(ctx context.Contex
 
 	if c.kind == enumspb.TASK_QUEUE_KIND_STICKY {
 		// In the sticky case we don't redirect, but we may kick off this worker if there's a newer one.
-		err := checkVersionForStickyAdd(data, buildId)
-		return c.taskQueueID, userDataChanged, err
+		unknownBuild, err := checkVersionForStickyAdd(data, buildId)
+		if err != nil {
+			return nil, nil, err
```
Comment (bergundy) [inline service/matching/task_queue_manager.go:823]: Should we return `userDataChanged` here?

File: service/matching/task_queue_manager.go [hunk 10]
```diff
@@ -809,20 +818,52 @@ func (c *taskQueueManagerImpl) RedirectToVersionedQueueForAdd(ctx context.Contex
...
+		// Send rpc to root partition to persist the unknown build id before we return success.
+		_, err = c.matchingClient.UpdateWorkerBuildIdCompatibility(ctx, &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+			NamespaceId: c.taskQueueID.namespaceID.String(),
+			TaskQueue:   c.taskQueueID.Root().FullName(),
+			Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+				PersistUnknownBuildId: buildId,
+			},
+		})
+		if err != nil {
+			return nil, nil, err
```
Comment (bergundy) [inline service/matching/task_queue_manager.go:851]: Do we need `userDataChanged` here?

**Review by bergundy — APPROVED** [review]
LGTM, had a few minor comments

File: service/matching/matching_engine_test.go [hunk 11]
```diff
@@ -2322,6 +2322,123 @@ func (s *matchingEngineSuite) TestAddActivityTask_ForVersionedWorkflows_Silently
 	s.Require().NoError(err)
 }
 
+func (s *matchingEngineSuite) TestPersistUnknownBuildId_Poll() {
```
[the remark this answers is not in the document]
Comment (dnr) [reply service/matching/matching_engine_test.go:2325]: true

File: service/matching/task_queue_manager.go (hunk 9 above)
Comment (dnr) [reply service/matching/task_queue_manager.go:823]: I don't think we need to. None of the callers look at it if err is non-nil. As much as possible, I'd like to keep to the pattern of not returning both error and values. (I know I break that rule in the dlq PR, but it's only for errUserDataDisabled and we can remove it all when we get rid of the kill switch)
